File: src/telegram/mediaDescription.ts
import { Api } from "telegram";
import { isVoiceMessage } from "@/telegram/utils";

export function formatMediaMessage(message: Api.Message): string | null {
	const media = message.media;
	if (!media) return null;

	// Photo (grouped photos are handled separately in formatMessagesWithUsernames)
	if (media instanceof Api.MessageMediaPhoto) {
		return "*photo*";
	}

	// Document-based media
	if (media instanceof Api.MessageMediaDocument) {
		const document = media.document;
		if (document instanceof Api.Document) {
			const isSticker = document.attributes?.some(
				(attr) => attr instanceof Api.DocumentAttributeSticker,
			);
			if (isSticker) {
				return "*sticker*";
			}

			// Voice message
			if (isVoiceMessage(message)) {
				return "*voice message*";
			}

			// Video
			const isVideo = document.attributes?.some(
				(attr) => attr instanceof Api.DocumentAttributeVideo,
			);
			if (isVideo) {
				return "*video message*";
			}

			// Audio (music)
			const isAudio = document.attributes?.some(
				(attr) =>
					attr instanceof Api.DocumentAttributeAudio &&
					!(attr as Api.DocumentAttributeAudio).voice,
			);
			if (isAudio) {
				return "*audio file*";
			}

			// Other documents (files)
			const mimeType = document.mimeType;
			if (mimeType) {
				let fileType = "file";
				const fileNameAttr = document.attributes?.find(
					(attr) => attr instanceof Api.DocumentAttributeFilename,
				);

				if (fileNameAttr instanceof Api.DocumentAttributeFilename) {
					const ext = fileNameAttr.fileName.split(".").pop()?.toLowerCase();
					if (ext) {
						fileType = `${ext} file`;
					}
				} else {
					const mimeParts = mimeType.split("/");
					if (mimeParts.length === 2) {
						fileType = `${mimeParts[1]} file`;
					}
				}

				return `*${fileType}*`;
			}

			return "*file*";
		}
	}

	// Contact
	if (media instanceof Api.MessageMediaContact) {
		return "*contact*";
	}

	// Location
	if (
		media instanceof Api.MessageMediaGeo ||
		media instanceof Api.MessageMediaVenue
	) {
		return "*location*";
	}

	// Poll
	if (media instanceof Api.MessageMediaPoll) {
		return "*poll*";
	}

	return "*media*";
}

File: src_py/telegram_utils/media_description.py
from telethon.tl import types

from src_py.telegram_utils.utils import is_voice_message


def format_media_message(message: types.Message) -> str | None:
    media = message.media
    if media is None:
        return None

    if isinstance(media, types.MessageMediaPhoto):
        return "*photo*"

    if isinstance(media, types.MessageMediaDocument):
        doc = media.document
        if isinstance(doc, types.Document):
            for attr in doc.attributes or []:
                if isinstance(attr, types.DocumentAttributeSticker):
                    return "*sticker*"

            if is_voice_message(message):
                return "*voice message*"

            for attr in doc.attributes or []:
                if isinstance(attr, types.DocumentAttributeVideo):
                    return "*video message*"

            for attr in doc.attributes or []:
                if isinstance(attr, types.DocumentAttributeAudio) and not attr.voice:
                    return "*audio file*"

            mime_type = doc.mime_type
            if mime_type:
                file_type = "file"
                for attr in doc.attributes or []:
                    if isinstance(attr, types.DocumentAttributeFilename):
                        ext = attr.file_name.rsplit(".", 1)[-1].lower() if "." in attr.file_name else ""
                        if ext:
                            file_type = f"{ext} file"
                        break
                else:
                    parts = mime_type.split("/")
                    if len(parts) == 2:
                        file_type = f"{parts[1]} file"

                return f"*{file_type}*"

            return "*file*"

    if isinstance(media, types.MessageMediaContact):
        return "*contact*"

    if isinstance(media, (types.MessageMediaGeo, types.MessageMediaVenue)):
        return "*location*"

    if isinstance(media, types.MessageMediaPoll):
        return "*poll*"

    return "*media*"
